### PR TITLE
Support the buffer protocol

### LINKED ISSUE #989: Support for the buffer protocol
It would be very useful if Numba could handle objects that support the buffer protocol in the same way we handle Numpy arrays.

### PR DESCRIPTION
**Pull request by pitrou**
Fix #989: support for the buffer protocol.
This supports simple scalar types. The more complex buffers are not handled (we could reuse Numpy's internal parsing routine for that).
Also, the memoryview type and its attributes are now supported, as well as the "itemsize" attribute of Numpy arrays.

### CODE REVIEW

File: numba/pythonapi.py [hunk 1]
```diff
@@ -778,76 +805,51 @@ def serialize_object(self, obj):
...
             zero = Constant.null(istrue.type)
-            return builder.icmp(lc.ICMP_NE, istrue, zero)
+            val = builder.icmp(lc.ICMP_NE, istrue, zero)
+            return NativeValue(val, is_error=c_api_error())
+
+        elif isinstance(typ, types.Integer):
+            val = self.to_native_int(obj, typ)
+            return NativeValue(val, is_error=c_api_error())
 
         elif typ in types.unsigned_domain:
```
Comment (gmarkall) [inline numba/pythonapi.py]: I think these checks for `typ` in unsigned/signed domain aren't doing anything - as far as I can see, the types that are members of these frozen sets will all be handled by the `isinstance(typ, types.Integer)` branch above.

File: numba/pythonapi.py [hunk 2]
```diff
@@ -868,35 +871,42 @@ def to_native_value(self, obj, typ):
...
-            return val
+            return NativeValue(val, is_error=c_api_error())
+
+        elif isinstance(typ, types.Record):
+            buf = self.alloca_buffer()
+            ptr = self.extract_record_data(obj, buf)
+
+            with cgutils.if_unlikely(self.builder,
+                                     cgutils.is_null(self.builder, ptr)):
+                # XXX
```
Comment (gmarkall) [inline numba/pythonapi.py]: Is this a marker for setting an exception instead of returning the null value as is done at the moment?

Comment (pitrou) [inline numba/pythonapi.py]: Yes. I should make this more explicit.

File: numba/pythonapi.py (hunk 1 above)
Comment (pitrou) [inline numba/pythonapi.py]: Oh, right. That's an oversight.

File: numba/targets/arrayobj.py [hunk 3]
```diff
@@ -583,8 +587,50 @@ def array_ndim(context, builder, typ, value):
 def array_size(context, builder, typ, value):
     arrayty = make_array(typ)
     array = arrayty(context, builder, value)
+    return array.nitems
+
+
+@builtin_attr
+@impl_attribute(types.Kind(types.Array), "itemsize", types.intp)
+@impl_attribute(types.Kind(types.MemoryView), "itemsize", types.intp)
+def array_ndim(context, builder, typ, value):
```
Comment (gmarkall) [inline numba/targets/arrayobj.py]: should this be `array_itemsize`?

Comment (gmarkall) [inline numba/targets/arrayobj.py]: (Also similar for "contiguous", "c_contiguous", "f_contiguous" and "readonly" below

File: numba/pythonapi.py [hunk 4]
```diff
@@ -868,35 +857,39 @@ def to_native_value(self, obj, typ):
...
+            buf = self.alloca_buffer()
+            ptr = self.extract_record_data(obj, buf)
+            is_error = cgutils.is_null(self.builder, ptr)
+
+            ltyp = self.context.get_value_type(typ)
+            val = builder.bitcast(ptr, ltyp)
+
+            def cleanup():
+                self.release_buffer(buf)
+            return NativeValue(val, cleanup=cleanup)
```
Comment (gmarkall) [inline numba/pythonapi.py]: Should this also have `is_error=is_error`?

Comment (pitrou) [inline numba/pythonapi.py]: Oops, yes, my bad...

**Pull request merged**